### PR TITLE
Add support for soft serial to ATmega32U2

### PR DESCRIPTION
**Pull request by takai**
## Description

serial.c now support ATmega32U4 only. However, ATmega32U2 has a similar feature, and it's not difficult to add ATmega32U2. This change adds soft serial support to the ATmega32U2.

Thanks to @mtei for his help in sending this pull request. The most of the code came from his ideas.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [X] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by tzarc — APPROVED** [review]

**Comment by tzarc** [comment]
This will need a rebase onto `develop`.

**Comment by takai** [comment]
> This will need a rebase onto develop.

Thanks. The commit is moved to on the top of develop branch.

File: drivers/avr/serial.c [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 #ifdef SOFT_SERIAL_PIN
 
-#    ifdef __AVR_ATmega32U4__
+#    if defined(__AVR_ATmega32U4__) || defined(__AVR_ATmega32U2__)
```
Comment (fauxpark) [inline drivers/avr/serial.c:23]:
```suggestion
#    if defined(__AVR_ATmega16U2__) || defined(__AVR_ATmega32U2__) || defined(__AVR_ATmega16U4__) || defined(__AVR_ATmega32U4__)
```

File: drivers/avr/serial.c [hunk 2]
```diff
@@ -52,7 +52,7 @@
 #                define EICRx_BIT (~(_BV(ISC30) | _BV(ISC31)))
 #                define SERIAL_PIN_INTERRUPT INT3_vect
 #            endif
-#        elif SOFT_SERIAL_PIN == E6
+#        elif defined(__AVR_ATmega32U4__) && SOFT_SERIAL_PIN == E6
```
Comment (fauxpark) [inline drivers/avr/serial.c:55]:
```suggestion
#        elif defined(__AVR_ATmega16U4__) || defined(__AVR_ATmega32U4__) && SOFT_SERIAL_PIN == E6
```

File: drivers/avr/serial.c [hunk 3]
```diff
@@ -61,7 +61,7 @@
 #        endif
 
 #    else
-#        error serial.c now support ATmega32U4 only
+#        error serial.c now support ATmega32U4 and ATmega32U2 only
```
Comment (fauxpark) [inline drivers/avr/serial.c:64]:
```suggestion
#        error serial.c currently only supports ATmegaxxU2 and ATmegaxxU4
```

File: drivers/avr/serial.c [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 #ifdef SOFT_SERIAL_PIN
 
-#    ifdef __AVR_ATmega32U4__
+#    if defined(__AVR_ATmega32U4__) || defined(__AVR_ATmega32U2__)
 // if using ATmega32U4 I2C, can not use PD0 and PD1 in soft serial.
 #        ifdef USE_AVR_I2C
 #            if SOFT_SERIAL_PIN == D0 || SOFT_SERIAL_PIN == D1
```
Comment (fauxpark) [inline drivers/avr/serial.c:26]: The line below this should be changed to `ATmegaxxU4`.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: drivers/avr/serial.c (hunk 2 above)
Comment (mtei) [reply drivers/avr/serial.c]: `A || B && C`  is equivalent to `A || (B && C)`
This should be `(A || B) && C`

File: drivers/avr/serial.c (hunk 4 above)
Comment (takai) [reply drivers/avr/serial.c:26]: Fixed on fdb63faf70ef768193b5207b432641714dda6349

File: drivers/avr/serial.c [hunk 5]
```diff
@@ -52,7 +52,7 @@
 #                define EICRx_BIT (~(_BV(ISC30) | _BV(ISC31)))
 #                define SERIAL_PIN_INTERRUPT INT3_vect
 #            endif
-#        elif SOFT_SERIAL_PIN == E6
+#        elif defined(__AVR_ATmega16U4__) || defined(__AVR_ATmega32U4__) && SOFT_SERIAL_PIN == E6
```
Comment (fauxpark) [inline drivers/avr/serial.c:55]:
```suggestion
#        elif (defined(__AVR_ATmega16U4__) || defined(__AVR_ATmega32U4__)) && SOFT_SERIAL_PIN == E6
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: drivers/avr/serial.c (hunk 2 above)
Comment (fauxpark) [reply drivers/avr/serial.c]: Oops. Got sidetracked by something else before I hit the button on this one, completely forgot to add those.

**Review by fauxpark — APPROVED** [review]

**Review by mtei — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**